spi.cpp

```diff
@@ -41,16 +41,19 @@ static uint32_t writeCounter = 0;
   TOGGLE_CHIP_SELECT_LINE(); \
   DEBUG_PRINT_WRITTEN_BYTE(w); \
   } while(0)
-
+  
 int mem_fd = -1;
 volatile void *bcm2835 = 0;
 volatile GPIORegisterFile *gpio = 0;
 volatile SPIRegisterFile *spi = 0;
 
 // Points to the system timer register. N.B. spec sheet says this is two low and high parts, in an 32-bit aligned (but not 64-bit aligned) address. Profiling shows
 // that Pi 3 Model B does allow reading this as a u64 load, and even when unaligned, it is around 30% faster to do so compared to loading in parts "lo | (hi << 32)".
+#if __aarch64__
+volatile uint32_t *systemTimerRegister = 0;
+#else
 volatile uint64_t *systemTimerRegister = 0;
-
+#endif

```

Comment:
I wonder if instead of the above, it would work to do

```cpp
#if __aarch64__
typedef uint64_t __attribute__((aligned(4))) timer_uint64_t;
#else
typedef uint64_t timer_uint64_t;
#endif

volatile timer_uint64_t *systemTimerRegister = 0;
```

This will tell the compiler that `systemTimerRegister` points to something that is only 32-bit aligned, and it should then automatically generate the appropriate 32-bit load and stores on its own.

After that, the other `#ifdef __aarch64__`s would not be necessary below?

Reply:
I've applied your patch on the following 64bit OS.
"Linux raspberrypi 5.15.84-v8+ #1613 SMP PREEMPT Thu Jan 5 12:03:08 GMT 2023 aarch64 GNU/Linux"

But it make same bus error.

bcm_host_get_peripheral_address: 0x3f000000, bcm_host_get_peripheral_size: 16777216, bcm_host_get_sdram_address: 0xc0000000
BCM core speed: current: 400000000hz, max turbo: 400000000hz. SPI CDIV: 6, SPI max frequency: 66666667hz
Initializing display
bcm2835 library version: 10071 (0x00002757)
Creating SPI task thread
InitSPI done
Bus error

Reply:
Oh, bummer. That would have been sweet if that had worked.